contracts/YieldEulerLever.sol

```diff
@@ -0,0 +1,390 @@
+    ///     the function will revert. Used to prevent slippage.
+    function borrow(
+        bytes6 ilkId,
+        bytes6 seriesId,
+        bytes12 vaultId,
+        uint256 borrowAmount,
+        uint256 fee,
+        uint128 baseAmount,
+        uint256 minCollateral
+    ) internal {
+        baseAmount += uint128(borrowAmount - fee);
+        // Deposit to get Euler token in return which would be used to payback flashloan
+        // Get the eToken address using the markets module:
+
+        IEulerEToken eToken = IEulerEToken(
+            markets.underlyingToEToken(ilkToAsset[ilkId])
+        );
+
+        eToken.deposit(0, borrowAmount - fee);
+
+        uint256 eBalance = IERC20(address(eToken)).balanceOf(address(this));
+
+        IERC20(address(eToken)).transfer(address(ladle.joins(ilkId)), eBalance);
+
+        _pourAndSell(vaultId, eBalance, borrowAmount - fee, seriesId);
+    }
+
+    /// @dev Additional function to get over stack too deep
+    /// @param vaultId VaultId
+    /// @param fCashAmount Amount of collateral
+    /// @param borrowAmount Amount being borrowed
+    /// @param seriesId SeriesId being
+    function _pourAndSell(
+        bytes12 vaultId,
+        uint256 fCashAmount,
+        uint256 borrowAmount,
+        bytes6 seriesId
+    ) internal {
+        IPool pool = IPool(ladle.pools(seriesId));
+        ladle.pour(
+            vaultId,
+            address(pool),
+            int128(uint128(fCashAmount)),
+            int128(pool.buyBasePreview(uint128(borrowAmount)))
+        );
+        pool.sellFYToken(address(this), 0);
+    }
+
+    /// @notice divest a position.
+    ///
+    ///     If pre maturity, borrow liquidity tokens to repay `art` debt and
+    ///     take `ink` collateral.
+    ///
+    ///     If post maturity, borrow USDC/DAI to pay off the debt directly.
+    ///
+    ///     This function will take the vault from you using `Giver`, so make
+    ///     sure you have given it permission to do that.
+    /// @param ilkId The id of the ilk being divested.
+    /// @param seriesId The seriesId corresponding to the vault.
+    /// @param vaultId The vault to use.
+    /// @param ink The amount of collateral to recover.
+    /// @param art The debt to repay.
+    /// @dev It is more gas efficient to let the user supply the `seriesId`,
+    ///     but it should match the pool.
+    function divest(
+        bytes6 ilkId,
+        bytes6 seriesId,
+        bytes12 vaultId,
+        uint128 ink,
+        uint128 art
+    ) external {
+        // Test that the caller is the owner of the vault.
+        // This is important as we will take the vault from the user.
+        require(cauldron.vaults(vaultId).owner == msg.sender);
+
+        // Give the vault to the contract
+        giver.seize(vaultId, address(this));
+
+        // Check if we're pre or post maturity.
+        if (uint32(block.timestamp) < cauldron.series(seriesId).maturity) {
```

Comment:
We might want to have this contract revert if `block.timestamp > type(uint32).max`

Reply:
When `block.timestamp == type(uint32).max`, `uint32(block.timestamp)` will overflow and return 00:00/1/1/1970, and after that L279 will thin that we are pre-maturity for all series, including those that matured far in the past.